src/Aquarius.Client/Samples/Client/FileUploader.cs

```diff
@@ -0,0 +1,110 @@
+        {
+        }
+
+        public FileUploader(IRestClient restClient, string authorizationHeader, string userAgent)
+        {
+            _restClient = restClient;
+            _restClient.AddHeader(SamplesClient.AuthorizationHeaderKey, authorizationHeader);
+            _restClient.AddHeader("User-Agent", userAgent);
+        }
+
+        public TResponse PostFileWithRequest<TResponse>(
+            string relativeOrAbsoluteUri,
+            Stream contentToUpload,
+            string uploadedFileName,
+            IReturn<TResponse> requestDto)
+        {
+            return _restClient.Post<TResponse>(
+                ComposePostUrlWithQueryParams(relativeOrAbsoluteUri, requestDto),
+                CreateMultipartContent(relativeOrAbsoluteUri, contentToUpload, uploadedFileName));
+        }
+
+        private string ComposePostUrlWithQueryParams<TResponse>(
+            string relativeOrAbsoluteUri,
+            IReturn<TResponse> requestDto)
+        {
+            var queryString = QueryStringSerializer.SerializeToString(requestDto);
+
+            var uriBuilder = new UriBuilder(relativeOrAbsoluteUri) {Query = queryString};
+
+            return uriBuilder.ToString();
+        }
+
+        private MultipartContent CreateMultipartContent(
+            string relativeOrAbsoluteUri,
+            Stream contentToUpload,
+            string uploadedFilename)
+        {
+            return IsImportServiceUpload(relativeOrAbsoluteUri)
+                ? CreateImportServiceUploadContent(contentToUpload, uploadedFilename)
+                : CreateFineUploaderContent(contentToUpload, uploadedFilename);
+        }
+
+        internal static bool IsImportServiceUpload(string relativeOrAbsoluteUri)
+        {
+            // TODO: We need a better way to distinguish between file upload request types
+            const string importServiceRoute = "/services/import/";
+
+            return relativeOrAbsoluteUri.ToLowerInvariant().Contains(importServiceRoute);
+        }
+
+        private static MultipartFormDataContent CreateImportServiceUploadContent(Stream contentToUpload, string uploadedFilename)
+        {
+            return CreateMultipartContent(contentToUpload, uploadedFilename, "file");
+        }
+
+        private static MultipartFormDataContent CreateFineUploaderContent(Stream contentToUpload, string uploadedFilename)
+        {
+            return CreateMultipartContent(contentToUpload, uploadedFilename, "qqfile", (content, contentBytes) =>
+            {
+                AddFormDataContent(content, new StringContent(uploadedFilename), "qqfilename");
+                AddFormDataContent(content, new StringContent(Guid.NewGuid().ToString("D")), "qquuid");
+                AddFormDataContent(content, new StringContent(contentBytes.Length.ToString(CultureInfo.InvariantCulture)), "qqtotalfilesize");
+            });
```

Comment:
This is pretty gross. Perhaps we should do something on the Samples API so public consumers can use more sensible names.